### DISCUSSION TITLE
Structure suggestion

### DISCUSSION THREAD

**Discussion by MaartenUreel** [wagtail/wagtail#10396] [General]
I'm coming from a pure Django-environment and I have stumbled upon Wagtail, and I like it :)

It's a bit unclear though when to use a snippet and when to use a page.

In essence the website is for an organization that organizes a few events per year, every event has a few artists, and very sometimes these are recurring.

Every edition also has partners (sponsors) which are mostly recurring.

So I have:
- Edition: Page
- Artist: Page (every artist gets a page where it is also shown when they were a guets)
- Partner: snippet
- EditionPartnership (M2M model between, Partners & Editions)
- EditionArtist (M2M model between, Artists & Editions)

Does this all make sense or am I looking at it incorrectly?

I started with the 5rc version since by the time I'll be done, it should be stable 😬

**Comment by lb-** [comment]
I'd recommend this approach - if something needs its own URL - a Page is probably the right model.

Snippets are great when you need a model that represents some isolated, repeated content that is not suitable as a Page.

For example, let's say you have some common parts of events like a Venue or a Speaker. These would be great candidates for Snippets as they can be used across multiple events.

**Reply by laymonage** [reply]
To add to this: there may be cases where you want the Venue or Speaker to have their own pages as well.

If this is the case, you can either:
- Use a Page model with a foreign key to the snippet model. Common fields are stored on the snippet model, while page-specific data is stored on the page model. With this:
  - You can have fields that are only used on the snippet's page (not the snippet itself). Thus, when the snippet is included elsewhere, the page-specific data isn't included with it. This could save some performance, especially if you use a `StreamField` with many blocks on the page that aren't used anywhere else.
  - The snippet models are not tied to Wagtail's tree structure. This could be useful if you'd like to use/export them outside of Wagtail context (e.g. third-party API integrations).

or:
- Make them as Page models instead of snippets. With this:
  - All fields will be included when you use the model anywhere. Could be useful, could be not.
  - The models are tied to Wagtail's tree structure. It's less flexible if you want to work with the data outside of Wagtail context.

The performance thing can probably be worked around using Django ORM's `only()`/`defer()`, but you need to remember to use them.

**Reply by MaartenUreel** [reply]
For partners (sponsors) I'm pretty sure to use a snippet. We just want to display them on every page and they don't need their own page.

However, artists are trickier. An artist could join multiple editions.

So I was thinking:

- Create ArtistArchivePage model
- Create ArtistPage models, children of ArtistArchivePage
- Create EditionPage model
- An EditionPage refers to ArtistPage records, but they are not child of that edition.

Would this be a good approach? To me it's a bit akward that I would have an instance of ArtistArchivePage just for the purpose of storing ArtistPage's under it, but I'm new to Wagtail so maybe this is what is intended :)

**Reply by MaartenUreel** [reply]
Actually I think I found a better way.

- Artist remains a snippet, contains all essential data (origin, some pictures, FB page,..)
- Under Edition, there are EditionArtist pages which have FK to the Artist snippet. This way, e.g. edition 2023 has a subpage for every artist.
- I create an EditionIndex page, which lists all past editions (and the current one, doesn't matter).

My only concern is that duplicate content could exist, e.g.:
/2023/artist/some_dude could have almost the same content as /2020/artist/some_dude.

There would be some unique content (e.g. pictures from that edition), but the description etc about the artist would be duplicate. And that could be bad for SEO, right? But that might be a discussion for another board :)
